agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/ConnectionString.java

```diff
@@ -52,16 +53,27 @@ public static void parseInto(String connectionString, TelemetryClient telemetryC
   }
 
   public static void updateStatsbeatConnectionString(
-      String ikey, String endpoint, TelemetryClient config)
+      @Nullable String ikey, @Nullable String endpoint, TelemetryClient telemetryClient)
       throws InvalidConnectionStringException {
     if (Strings.isNullOrEmpty(ikey)) {
       logger.warn("Missing Statsbeat '" + Keywords.INSTRUMENTATION_KEY + "'");
     }
 
-    config.setStatsbeatInstrumentationKey(ikey);
+    // if customer is in EU region and their statsbeat config is not in EU region, customer is
+    // responsible for breaking the EU data boundary violation.
+    // Statsbeat config setting has the highest precedence.
+    if (ikey == null || ikey.isEmpty()) {
+      StatsbeatConnectionString.InstrumentationKeyEndpointPair pair =
+          StatsbeatConnectionString.getInstrumentationKeyAndEndpointPair(
+              telemetryClient.getEndpointProvider().getIngestionEndpoint().toString());
+      ikey = pair.instrumentationKey;
+      endpoint = pair.endpoint;
+    }
```

Comment:
can we move this code to `ConfigurationBuilder`, near the end of `create`? that seems maybe better place for configuration fallbacks, then we can just pass the configured values around and other components don't need to worry about it